### DISCUSSION TITLE
Logger MDC auto-instrumentation with opentelemetry-spring-boot-starter

### DISCUSSION THREAD

**Discussion by birkoni5** [open-telemetry/opentelemetry-java-instrumentation#7653] [Q&A] [answered]
Hello,

I have a problem with a correlation between logs and traces, log statements don't contain any tracing information when  [`opentelemetry-spring-boot-starter`](https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/17e1e22072fc7c0b5ae43f54cf8af668da4cb730/instrumentation/spring/starters/spring-boot-starter/README.md) `1.22.0-alpha` is used with Logback `1.2.11` and Spring Boot `2.7.2`. Is it possible to inject `trace_id` and `span_id` into MDC to be included in produced log statements?  

Any suggestion on how tracing information can be appended to each log statement is appreciated.

Thank you.

**Comment by mateuszrzeszutek** [comment, marked as answer]
Hey @birkoni5 ,

Unfortunately the spring boot starter is missing this feature right now; it can be added very easily though. You'll have to include [one more dependency](https://github.com/open-telemetry/opentelemetry-java-instrumentation/tree/main/instrumentation/logback/logback-mdc-1.0/library), and then configure your logging pattern to include the `trace_id`/`span_id` information.

**Comment by trajano** [comment]
It appears to be working with Spring Boot 3.1.x with the current instrumentation and works with the Logstash JSON encoder as well

> {"@timestamp":"2023-11-07T23:55:37.814277259-05:00","@version":"1","message":"whomami 2 requested","logger_name":"net.trajano.simpleauth.RootController","thread_name":"lettuce-nioEventLoop-5-1","level":"INFO","level_value":20000,"trace_id":"6c896aab0066650b36cd5a09f8188f69","trace_flags":"00","span_id":"7e1d76ca2865f464"}

**Comment by antechrestos** [comment]
I dunno whether it is appropriate but I instanciated the initializer as follow

```java
public class OtelLogbackMdcAppenderApplicationInitializer implements GenericApplicationListener {
    @Override
    public boolean supportsEventType(ResolvableType eventType) {
        return eventType.isAssignableFrom(ApplicationEnvironmentPreparedEvent.class);
    }

    @Override
    public void onApplicationEvent(ApplicationEvent event) {
        if(event instanceof ApplicationEnvironmentPreparedEvent) {
            Logger logger = (Logger) LoggerFactory.getILoggerFactory().getLogger("ROOT");
            OpenTelemetryAppender openTelemetryAppender = new OpenTelemetryAppender();
            List<Appender<ILoggingEvent>> existingAppelders = new ArrayList<>();
            logger.iteratorForAppenders().forEachRemaining(existingAppelders::add);
            existingAppelders.forEach(logger::detachAppender);
            AtomicBoolean appenderAddedToOpenTelemetry = new AtomicBoolean();
            existingAppelders.forEach(existingAppender -> {
                if(existingAppender instanceof OpenTelemetryAppender
                || existingAppender instanceof io.opentelemetry.instrumentation.logback.appender.v1_0.OpenTelemetryAppender) {
                    logger.addAppender(existingAppender);
                } else {
                    openTelemetryAppender.addAppender(existingAppender);
                    appenderAddedToOpenTelemetry.set(true);
                }
            });
            if(appenderAddedToOpenTelemetry.get()) {
                openTelemetryAppender.start();
                logger.addAppender(openTelemetryAppender);
            }
        }
    }
}
```

and add it in a `spring.factories`file; this inspired from the `LogbackAppenderInstaller`..
